Lombiq.Npm.Targets.targets

```diff
@@ -63,13 +63,105 @@
            AlwaysCreate="true" />
   </Target>
 
+  <UsingTask TaskName="CopyVendors"
+             TaskFactory="RoslynCodeTaskFactory"
+             AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll" >
+    <ParameterGroup>
+      <ProjectPath ParameterType="System.String" Required="true" />
+      <ProjectFilePath ParameterType="System.String" Required="true" />
+      <LibraryPath ParameterType="System.String" Required="true" />
+    </ParameterGroup>
+    <Task>
+      <Using Namespace="System" />
+      <Using Namespace="System.IO" />
+      <Using Namespace="System.Linq" />
+      <Using Namespace="System.Xml.Linq" />
+      <Using Namespace="System.Xml.XPath" />
+      <Code Type="Fragment" Language="cs">
+        <![CDATA[
+        static void CopyDirectory(DirectoryInfo source, string target)
+        {
+            Directory.CreateDirectory(target);
+
+            foreach (var file in source.GetFiles())
+            {
+                file.CopyTo(Path.Combine(target, file.Name));
+            }
+
+            foreach (var subdirectory in source.GetDirectories())
+            {
+                CopyDirectory(subdirectory, Path.Combine(target, subdirectory.Name));
+            }
+        }
+
+        static void FindVendors(string projectPath, string projectFilePath, Action<string> error, Action<string> warning)
+        {
+            var document = XDocument.Load(projectFilePath);
+            var vendors = document.Root.XPathSelectElements("//ItemGroup/Vendor");
```

Comment:
Isn't there any high-level way of reading these instead of processing the XML? Because people can e.g. use MSBuild properties in `Include` which should be resolved by MSBuild.

Reply:
I will check if there already is a csproj parser, because that would be less maintenance. But I don't think we should support a broad MSBuild feature set unless it's super trivial to utilize, because it doesn't make much sense for this use-case. We really just want a list of package paths to copy from, using MSBuild properties would hurt readability and we don't do it in other cases such as ProjectReference either.

Reply:
I wouldn't complicate this either, but us processing the XML directly seems like a can of worms, and I suspect that eventually, we'll bump into issues.

I have a faint memory that at one point, we used something that provided the feature of "give me the csproj with everything hydrated, as MSBuild would resolve it". I suspected it to be related to https://github.com/Lombiq/GitHub-Actions/blob/dev/.github/actions/publish-nuget/Add-SourceLinkPackage.ps1, but there's nothing like this currently, nor do I see it in the history of it (but the .NET CLI definitely does something of this effect at least in `dotnet add`).

But if there's nothing straightforward, then let's leave it.

Reply:
Unfortunately the [Microsoft.Build](https://www.nuget.org/packages/Microsoft.Build) package which contains the classes to [evaluate](https://github.com/dotnet/msbuild/tree/695b1ef88c242c19a95c9f6a504a2376661762c2/src/Build#:~:text=for%20evaluating%20MSBuild%20projects%2C) or create MSBuild projects does not support the .NETStandard 2.0 framework which is a hard requirement for MSBuild tasks. Ironic.

By the way I'm not worried about the kind of problems you mentioned, because here we are really just using `<ItemsGroup>` as a way to store arbitrary user data in the project file.

Reply:
I have added a warning [to the readme](https://github.com/Lombiq/NPM-Targets/blob/acc3dcc547bf7f077ce05660e40a70b6646c0683/Readme.md#copying-vendor-libraries-into-wwwroot) to explain the limitations.

Reply:
OK then!